pkg/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package ratelimit
+
+import (
+	"time"
+
+	"go.uber.org/ratelimit"
+)
+
+type Event struct {
+	Timestamp time.Time
+	Tag       int
+}
+
+type RateLimiter struct {
+	outputCh chan Event
+	C        <-chan Event
+	src      <-chan Event
+	rt       ratelimit.Limiter
+	doneCh   chan struct{}
+}
+
+func NewWithEPS(numberOfEvents int64, timeUnit time.Duration, src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	opt := ratelimit.Per(timeUnit)
+	rt.rt = ratelimit.New(int(numberOfEvents), opt)
+	return rt
+}
+
+func NewUnlimited(src <-chan Event) *RateLimiter {
+	rt := newCommon(src)
+	rt.rt = ratelimit.NewUnlimited()
+	return rt
+}
+
+func newCommon(src <-chan Event) *RateLimiter {
+	rt := RateLimiter{
+		src:      src,
+		outputCh: make(chan Event),
+		doneCh:   make(chan struct{}),
+	}
+	rt.C = rt.outputCh
+	return &rt
+}
+
+func (rt *RateLimiter) Run() {
```

Comment:
Please correct me if I am wrong, but I do not see any blocking loops under the `Run` meaning, it will exit straight forward after the call and also will exit the go routine.

Reply:
Yes, you are right, there is no blocking loops. I would have called this method "Start" instead of "Run" but I called "Run" to follow the names I have seen  in the RTE (ResourceObserver for example )

Reply:
I thought we should always run this loop and does not exit straightforward after the call.
IMHO we should change it to
```
func (rt *RateLimiter) Run() {
for {
			select {
			case event := <-rt.src:
				rt.rt.Take()
				rt.outputCh <- event
			case <-rt.doneCh:
				return
			}
		}
}
```
and under the `Execute` call it via
`go Run(...)`

Reply:
Yep I have nothing against that change ... I just did it that way to keep some homogeneity with the current code.
If we think this is a better way to do it, just to left concurrency in the caller hands, maybe we should also change ResourceObserver to behave the same way.

Reply:
yes, we can use some cleanup in this area for this project. Let's pause for a sec while we evaluate a cleanup PR before we move forward.

Reply:
My main concern is that we exit the Run straight forward after the call and close the blocking goroutine

Reply:
But I can see we already have https://github.com/k8stopologyawareschedwg/resource-topology-exporter/pull/94/files
so it ok for me.